### ISSUE TITLE
Some doubts about batch size

### ISSUE THREAD

**Issue by Zoisaang** [volcengine/verl#1524]
Here are the settings for GRPO in the example codes ([examples/grpo_trainer/run_deepseek7b_llm.sh](https://github.com/volcengine/verl/blob/main/examples/grpo_trainer/run_deepseek7b_llm.sh))：

data.train_batch_size=1024
actor_rollout_ref.actor.ppo_mini_batch_size=256
actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=80
actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=160
actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=160
actor_rollout_ref.rollout.n=5

I’d like to know whether data.train_batch_size = 1024 refers to #prompts, or #prompts * actor_rollout_ref.rollout.n. In other words，if batch data is 1024*5 or 1024?

actor_rollout_ref.actor.ppo_mini_batch_size=256, is it refers to #prompts or #prompts * actor_rollout_ref.rollout.n?

actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=80 cannot divide the actor_rollout_ref.actor.ppo_mini_batch_size=256 evenly — why is that? Is this a misconfiguration in the example code?

Does the value of actor_rollout_ref.actor.ppo_mini_batch_size=256 and actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu = 80 consider actor_rollout_ref.rollout.n=5？What I mean is that how do I calculate my training steps？Is 10245 / 256 or 1024 / 256？And how do I calculate accumulated steps？Is 256 / 80 or 2565/80？

🤗 I would appreciate it a lot if anyone could help me clarify these questions. 🤗 🌻

**Comment by 0x404** [comment]
Hi @Zoisaang, this is my understanding, hope this could help and please correct me if I am wrong:

1. In the config file, `data.train_batch_size` refers to the number of prompts. However, during the actual training process, the real training batch size equals `data.train_batch_size * actor_rollout_ref.rollout.n`.
2. In the config file, `actor.ppo_mini_batch_size` refers to the number of prompts. But in the verl code, it will be normalized to `actor.ppo_mini_batch_size * actor_rollout_ref.rollout.n`.
3. In this example case, `ppo_mini_batch_size` is set to 256, and it will be normalized to `actor.ppo_mini_batch_size * actor_rollout_ref.rollout.n`, which is 1280. This means, from a global view, each mini step will consume 1280 examples. From the local view (this config uses 1 node and 8 GPUs), one GPU should consume 1280/8 = 160 examples. And `ppo_micro_batch_size_per_gpu` is a local view value, which means `ppo_micro_batch_size_per_gpu` should be a value that evenly divides 160 examples.
4. One global step (step in the ver log) consumes `ppo_mini_batch_size * actor_rollout_ref.rollout.n` examples, which is 1280 examples in this case.

**Comment by Zoisaang** [comment]
> Hi [@Zoisaang](https://github.com/Zoisaang), this is my understanding, hope this could help and please correct me if I am wrong:
> 
> 1. In the config file, `data.train_batch_size` refers to the number of prompts. However, during the actual training process, the real training batch size equals `data.train_batch_size * actor_rollout_ref.rollout.n`.
> 2. In the config file, `actor.ppo_mini_batch_size` refers to the number of prompts. But in the verl code, it will be normalized to `actor.ppo_mini_batch_size * actor_rollout_ref.rollout.n`.
> 3. In this example case, `ppo_mini_batch_size` is set to 256, and it will be normalized to `actor.ppo_mini_batch_size * actor_rollout_ref.rollout.n`, which is 1280. This means, from a global view, each mini step will consume 1280 examples. From the local view (this config uses 1 node and 8 GPUs), one GPU should consume 1280/8 = 160 examples. And `ppo_micro_batch_size_per_gpu` is a local view value, which means `ppo_micro_batch_size_per_gpu` should be a value that evenly divides 160 examples.
> 4. One global step (step in the ver log) consumes `ppo_mini_batch_size * actor_rollout_ref.rollout.n` examples, which is 1280 examples in this case.

Hi @0x404 , thank you very much for your attention and valuable response. Based on what you mentioned, it seems that when setting $$data.train_batch_size$$ and $$actor_rollout_ref.actor.ppo_mini_batch_size$$, the target should be the number of prompts. However, when configuring $$actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu$$, we should take $$actor_rollout_ref.rollout.n into account$$ — is that correct? 

If so, then would the correct formula for accumulate_steps be:
$$accumulate_steps = (actor_rollout_ref.actor.ppo_mini_batch_size * actor_rollout_ref.rollout.n) / total_gpus / actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu$$

Is my understanding accurate?

**Comment by 0x404** [comment]
> is that correct?

Yes, I think you are right.

> If so, then would the correct formula for accumulate_steps be:
accumulate_steps = (actor_rollout_ref.actor.ppo_mini_batch_size * actor_rollout_ref.rollout.n) / total_gpus / actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu

Yes, you can refer to the code here: https://github.com/volcengine/verl/blob/771bd756b35f4766b1762a5e3626c104f20b4b4f/verl/workers/actor/dp_actor.py#L280

**Comment by Zoisaang** [comment]
> > is that correct?
> 
> Yes, I think you are right.
> 
> > If so, then would the correct formula for accumulate_steps be:
> > accumulate_steps = (actor_rollout_ref.actor.ppo_mini_batch_size * actor_rollout_ref.rollout.n) / total_gpus / actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu
> 
> Yes, you can refer to the code here:
> 
> [verl/verl/workers/actor/dp_actor.py](https://github.com/volcengine/verl/blob/771bd756b35f4766b1762a5e3626c104f20b4b4f/verl/workers/actor/dp_actor.py#L280)
> 
> Line 280 in [771bd75](/volcengine/verl/commit/771bd756b35f4766b1762a5e3626c104f20b4b4f)
> 
>  self.gradient_accumulation = self.config.ppo_mini_batch_size // self.config.ppo_micro_batch_size_per_gpu

Once again, thank you for your helpful response :)

**Comment by NICKQZL** [comment]
> Hi [@Zoisaang](https://github.com/Zoisaang), this is my understanding, hope this could help and please correct me if I am wrong:
> 
> 1. In the config file, `data.train_batch_size` refers to the number of prompts. However, during the actual training process, the real training batch size equals `data.train_batch_size * actor_rollout_ref.rollout.n`.
> 2. In the config file, `actor.ppo_mini_batch_size` refers to the number of prompts. But in the verl code, it will be normalized to `actor.ppo_mini_batch_size * actor_rollout_ref.rollout.n`.
> 3. In this example case, `ppo_mini_batch_size` is set to 256, and it will be normalized to `actor.ppo_mini_batch_size * actor_rollout_ref.rollout.n`, which is 1280. This means, from a global view, each mini step will consume 1280 examples. From the local view (this config uses 1 node and 8 GPUs), one GPU should consume 1280/8 = 160 examples. And `ppo_micro_batch_size_per_gpu` is a local view value, which means `ppo_micro_batch_size_per_gpu` should be a value that evenly divides 160 examples.
> 4. One global step (step in the ver log) consumes `ppo_mini_batch_size * actor_rollout_ref.rollout.n` examples, which is 1280 examples in this case.

Hello, i feel little bit confused with that since we already have parameter "mini_batch_size"(i.e. actor.ppo_mini_batch_size), why we still need parameter "data.train_batch_size" to differenciate the batch_num of prompts during rollout period and batch_num of prompts fed into optimizer?

**Comment by juls173** [comment]
> > Hi [@Zoisaang](https://github.com/Zoisaang), this is my understanding, hope this could help and please correct me if I am wrong:
> > 
> > 1. In the config file, `data.train_batch_size` refers to the number of prompts. However, during the actual training process, the real training batch size equals `data.train_batch_size * actor_rollout_ref.rollout.n`.
> > 2. In the config file, `actor.ppo_mini_batch_size` refers to the number of prompts. But in the verl code, it will be normalized to `actor.ppo_mini_batch_size * actor_rollout_ref.rollout.n`.
> > 3. In this example case, `ppo_mini_batch_size` is set to 256, and it will be normalized to `actor.ppo_mini_batch_size * actor_rollout_ref.rollout.n`, which is 1280. This means, from a global view, each mini step will consume 1280 examples. From the local view (this config uses 1 node and 8 GPUs), one GPU should consume 1280/8 = 160 examples. And `ppo_micro_batch_size_per_gpu` is a local view value, which means `ppo_micro_batch_size_per_gpu` should be a value that evenly divides 160 examples.
> > 4. One global step (step in the ver log) consumes `ppo_mini_batch_size * actor_rollout_ref.rollout.n` examples, which is 1280 examples in this case.
> 
> Hello, i feel little bit confused with that since we already have parameter "mini_batch_size"(i.e. actor.ppo_mini_batch_size), why we still need parameter "data.train_batch_size" to differenciate the batch_num of prompts during rollout period and batch_num of prompts fed into optimizer?

The train batch * rollout.n determines the amount of generations sampled from the model at the beginning of every step, and the mini batches (normalized by rollout.n) are then sampled from these generations. There are (train_batch_size / mini_batch_size) gradient updates for every step. For example, if your train_batch_size is set to 512 and mini_batch_size is set to 128, there will be 4 gradient updates before you generate samples again from your now-updated model, meaning that 3 of these 4 gradient updates were not on-policy (as they are based on data sampled from the model at the beginning of the step, which has already been updated). However, a benefit of increasing train_batch_size is that you have fewer overall steps (dataset size / train_batch_size), decreasing the overhead of switching between the inference engine and optimization.

**Comment by AragornHorse** [comment]
I have a different understanding, please correct me if I am wrong:
I think ```data.train_batch_size``` refers to the number of prompts, while ```actor_rollout_ref.actor.ppo_mini_batch_size``` refers to the number of samples, because I found the following code in the source code: 
```
# verl/verl/workers/actor/dp_actor.py
# Line 430
mini_batches = data.split(self.config.ppo_mini_batch_size)
```
I think this means it split all examples into batches, each size is ```actor_rollout_ref.actor.ppo_mini_batch_size```.
But I have another question: I noticed that in most recipes, the ```actor_rollout_ref.actor.ppo_mini_batch_size``` values are all very small, typically 32, and sometimes even 1. Why is that?

**Comment by jairuigou** [comment]
> while `actor_rollout_ref.actor.ppo_mini_batch_size` refers to the number of samples, because I found the following code in the source code

There is a normalization step here. It converts the prompt count into the total number of samples.
https://github.com/volcengine/verl/blob/ed8b39849f8b52f84d3ea4d6baffb0b33888b2dd/verl/workers/fsdp_workers.py#L242

**Comment by JayPritchet** [comment]
> > > Hi [@Zoisaang](https://github.com/Zoisaang), this is my understanding, hope this could help and please correct me if I am wrong:
> > > 
> > > 1. In the config file, `data.train_batch_size` refers to the number of prompts. However, during the actual training process, the real training batch size equals `data.train_batch_size * actor_rollout_ref.rollout.n`.
> > > 2. In the config file, `actor.ppo_mini_batch_size` refers to the number of prompts. But in the verl code, it will be normalized to `actor.ppo_mini_batch_size * actor_rollout_ref.rollout.n`.
> > > 3. In this example case, `ppo_mini_batch_size` is set to 256, and it will be normalized to `actor.ppo_mini_batch_size * actor_rollout_ref.rollout.n`, which is 1280. This means, from a global view, each mini step will consume 1280 examples. From the local view (this config uses 1 node and 8 GPUs), one GPU should consume 1280/8 = 160 examples. And `ppo_micro_batch_size_per_gpu` is a local view value, which means `ppo_micro_batch_size_per_gpu` should be a value that evenly divides 160 examples.
> > > 4. One global step (step in the ver log) consumes `ppo_mini_batch_size * actor_rollout_ref.rollout.n` examples, which is 1280 examples in this case.
> > 
> > 
> > Hello, i feel little bit confused with that since we already have parameter "mini_batch_size"(i.e. actor.ppo_mini_batch_size), why we still need parameter "data.train_batch_size" to differenciate the batch_num of prompts during rollout period and batch_num of prompts fed into optimizer?
> 
> The train batch * rollout.n determines the amount of generations sampled from the model at the beginning of every step, and the mini batches (normalized by rollout.n) are then sampled from these generations. There are (train_batch_size / mini_batch_size) gradient updates for every step. For example, if your train_batch_size is set to 512 and mini_batch_size is set to 128, there will be 4 gradient updates before you generate samples again from your now-updated model, meaning that 3 of these 4 gradient updates were not on-policy (as they are based on data sampled from the model at the beginning of the step, which has already been updated). However, a benefit of increasing train_batch_size is that you have fewer overall steps (dataset size / train_batch_size), decreasing the overhead of switching between the inference engine and optimization.

I agree with this. But there is a little bug in the description `There are (train_batch_size / mini_batch_size) gradient updates for every step.`. Actually, there are `train_batch_size / mini_batch_size*ppo_epochs` gradient updates. 
https://github.com/verl-project/verl/blob/5a506cc58c4cb6aaf5bff908726e55378eb42867/verl/utils/tensordict_utils.py#L607-L612

**Comment by Kirrito-k423** [comment]
中文：我根据当前代码和文档再补充一个澄清，避免这里的几个 batch-size 名字混在一起。

- `data.train_batch_size=1024` 是 prompt 数；在 `rollout.n=5` 时，一轮 rollout 会产生 `1024 * 5 = 5120` 条 responses/trajectories。当前文档也这样描述：`data.train_batch_size * actor_rollout_ref.rollout.n`。
- 对 actor 更新来说，当前 trainer 代码会把配置里的 `actor_rollout_ref.actor.ppo_mini_batch_size` 再乘以 `rollout.n` 后传给 actor update。因此可以把配置值 `256` 理解为 prompt 维度；内部用于 actor mini-batch 的 trajectory/sample 数是 `256 * 5 = 1280`。
- `ppo_micro_batch_size_per_gpu=80` 是每张 GPU 本地一次 forward/backward 的 micro batch 大小，不需要直接整除配置里的 `256`。应检查的是 normalized mini batch 在 actor data-parallel 维度切分后的本地大小能否被它整除。例如如果 actor 训练使用 8 个 data-parallel GPU，则本地 mini batch 是 `1280 / 8 = 160`，`160 / 80 = 2`，所以这里不是误配置。
- 每个 rollout batch 的 actor mini-batch 数量是 `(data.train_batch_size * rollout.n) / (ppo_mini_batch_size * rollout.n)`，也就是 `1024 / 256 = 4`；如果 `ppo_epochs > 1`，总的 actor update mini-iterations 还要再乘以 `ppo_epochs`。
- `actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu` 和 `actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu` 是 log-prob 计算的 per-GPU forward-only micro batch，主要影响显存和吞吐，不改变上述算法意义上的 train/mini batch 计数。

English: I re-checked the current code and docs and would summarize the batch-size semantics as follows.

- `data.train_batch_size=1024` is the number of prompts. With `rollout.n=5`, one rollout batch contains `1024 * 5 = 5120` responses/trajectories. This matches the current GRPO documentation.
- For actor updates, the current trainer multiplies the configured `actor_rollout_ref.actor.ppo_mini_batch_size` by `rollout.n` before passing it into the actor update path. So the configured value `256` can be read in prompt units, while the internal actor mini-batch contains `256 * 5 = 1280` trajectories/samples.
- `ppo_micro_batch_size_per_gpu=80` is a local per-GPU micro batch for one forward/backward chunk. It does not have to directly divide the configured `256`. The relevant check is whether the normalized mini-batch per actor data-parallel rank is divisible by it. For example, with 8 actor data-parallel GPUs, the local mini-batch is `1280 / 8 = 160`, and `160 / 80 = 2`, so this example is not a misconfiguration.
- The number of actor mini-batches per rollout batch is `(data.train_batch_size * rollout.n) / (ppo_mini_batch_size * rollout.n)`, i.e. `1024 / 256 = 4`; if `ppo_epochs > 1`, multiply that by `ppo_epochs` for total actor update mini-iterations.
- The rollout/ref `log_prob_micro_batch_size_per_gpu` settings are forward-only per-GPU chunks for log-prob computation. They mainly affect memory/throughput and do not change the algorithmic train/mini batch counts above.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
